server/src/main/java/org/candlepin/controller/ActivationKeyController.java

```diff
@@ -0,0 +1,246 @@
+/**
+ * Copyright (c) 2009 - 2019 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.controller;
+
+import org.candlepin.common.exceptions.AlreadyRegisteredException;
+import org.candlepin.common.exceptions.NotFoundException;
+import org.candlepin.common.exceptions.RuleValidationException;
```

Comment:
It is not clear to me how, when the controller throws `AlreadyRegisteredException` and `NotFoundException`, they are handled? It should be the responsibility of the resource to turn those into resource exceptions, but ActivationKeyResource doesn't do that. Or at least a more general mechanism should exist for that conversion.

Reply:
Exceptions are translated to appropriate responses using ExceptionMappers such as NotFoundExceptionMapper.